doc/design/service-discovery.md

```diff
@@ -0,0 +1,316 @@
+* Not just delta of service registry is transmitted, but complete service registry info
+* Fast data transport
+* Close to DDS mechanism
+* Could replace `InterfacePort`s
+
+Con:
+
+* New publisher in RouDi needed
+* Partially overlapping features with `InterfacePort`s
+  * Could the `InterfacePort`s be replace by alternative D?
+
+Note:
+
+* To avoid any out of memory issues with custom user-configured mempools or access rights problems, the publisher shall
+write into the `iceoryx_managment` segment (same applies for the introspection publishers)
+
+### Decision
+
+The **event-based notification alternative D** fulfills both requirements for a asychronous, one shot-polling approach as
+well as the event-based one. It can replace the `InterfacePort`s, which broadens the usage e.g. for gateways or even
+introspection 2.0. Furthermore, alternative D does not add any new non-trivial building blocks. The one-shot RPC call
+does not lead to a shared memory transmission compared to the alternative B (request/response feature usage) as the
+complete service registry is stored locally.
+
+### Code example
+
+#### Event-based notification: Alternative B
+
+```cpp
+auto& runtime = PoshRuntime::initRuntime("myApp");
+auto* interfacePortData = runtime.getMiddlewareInterface(capro::Interfaces::INTERNAL);
+
+InterfacePort caproSubscriber(interfacePortData);
+Listener myListner;
+
+void onDiscoveryUpdateCallback(InterfacePort* subscriber)
+{
+    subscriber->tryGetCaProMessage().and_then([](auto& caproMessage){
+        // Has any service that is relevant for me changed?
+        if(caproMessage.m_serviceDescription == myRelevantServiceDescription && someOtherCondition())
+        {
+            // call user-defined callback
+        }
+    });
+}
+
+myListner.attachEvent(caproSubscriber, DATA_RECEIVED, createNotificationCallback(onDiscoveryUpdateCallback));
+```
+
+#### Event-based notification: Alternative D
+
+```cpp
+class DiscoveryInfo
+{
+  public:
+
+    // Move all service-related methods from PoshRuntime to this class
+    cxx::expected<InstanceContainer, FindServiceError>
+    PoshRuntime::findService(const capro::ServiceDescription& serviceDescription) noexcept
+    {
+        // Update local service registry
+        get();
+
+        return filterFor(serviceDescription);
+    }
+
+    // @todo check if still needed
+    bool offerService(const capro::ServiceDescription& serviceDescription) noexcept;
+    void stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    ServiceRegistryTopic get() noexcept
+    {
+        m_subscriber.take().and_then([](auto& ServiceRegistryTopic){
+            // Update our local copy of the service registry
+            m_lastServiceRegistry = ServiceRegistryTopic;
+        }).or_else([](){
+            // No change in service registry
+        });
+        return m_lastServiceRegistry;
+    }
+
+    ServiceRegistryTopic getDelta() noexcept
```

Comment:
Computing the delta efficiently is important if it is used often. This may not be easy/possible with strings though but the general idea to compute the delta is useful/needed. We can also cache the result and use it if the state (`m_lastRegistry`) did not change, i.e. we did not receive any update, we do not need to recompute the delta.

Reply:
Still need to put some brainpower into how to do that efficiently. Let's iterate a couple of times over it.